[SYCL] Make nd_item stateless

bader opened the pull request:
The primary motivation for this change is to reduce the overhead of
nd_item object creation. The nd_item object is created at the beginning
of each "nd-range kernel". nd_item constructor initializes following
members:
- global item (global id, global range, global offset)
- local item (local id, local range)
- group (global range, local range, number of groups, group id)

Most applications do not use all these data, so initializing them is
unnecessary overhead. Due to compiler optimizations like aggressive
inlining, SROA and dead code elimination, the overhead can be avoided
in some cases.

This patch removes all nd_item members and uses SPIR-V intrinsics to get
access to the data we keep as nd_item members. This achieved though
following changes:
1. group class member functions async_workg_group_copy and wait_for 
    are inlined to nd_item class.
2. global and local item members are removed. The data obtained via
   SPIR-V instrinsics.

bader commented on sycl/include/CL/__spirv/spirv_vars.hpp:
```diff
@@ -47,13 +47,17 @@ __DPCPP_SYCL_EXTERNAL size_t __spirv_LocalInvocationId_x();
 __DPCPP_SYCL_EXTERNAL size_t __spirv_LocalInvocationId_y();
 __DPCPP_SYCL_EXTERNAL size_t __spirv_LocalInvocationId_z();
 
+__DPCPP_SYCL_EXTERNAL size_t __spirv_GlobalLinearId();
+__DPCPP_SYCL_EXTERNAL size_t __spirv_LocalInvocationIndex();
+
```
@intel/llvm-reviewers-cuda, I haven't implemented these functions for NVIDIA and AMDGPU targets.
What is the best way to do that?
I think I implement `__spirv_LocalInvocationIndex` myself. It can be done by adding implementation to libclc. 
`__spirv_GlobalLinearId` concerns me because it uses offset to compute linear id. I know that offset in SYCL is resolved in the compiler pass, but I'm not sure how offset is exposed to the OpenCL C language. I would appreciate your help with `__spirv_GlobalLinearId` implementation.

bader replied:
> I think that __spirv_GlobalLinearId maps to e.g. `__nvvm_read_ptx_sreg_tid_x` from
> 
> https://github.com/intel/llvm/blob/dc0848d35984d1d1d27739d8b73c8c275387bec9/clang/include/clang/Basic/BuiltinsNVPTX.def#L88
> 
> 
> which corresponds to the ptx:
> https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#special-registers-tid
> 
> I don't know about offset issues.

`__nvvm_read_ptx_sreg_tid_x` corresponds to `LocalIvocationId` built-in variable in SPIR-V (or `get_local_id` in OpenCL).

Let me switch to OpenCL terminology, which might be more convenient.
What I'm looking for is to express `get_local_linear_id` and `get_global_linear_id`.
I know how to compute `get_local_linear_id`. It's just get_local_id(2) * get_local_size(1) * get_local_size(0) + get_local_id(1) * get_local_size(0) + get_local_id(0); or using CUDA terms threadIdx.x + threadIdx.y * blockDim.x + threadIdx.z * blockDim.y * blockDim.x. Should be easy to implement libclc.

The problem is `get_global_linear_id`. For 1-dim case the value is get_global_id(0) - get_global_offset(0). There is "offset" concept in CUDA, so we added a new built-in recognized by SYCL compiler (`llvm.{amdgcn|nvvm}.implicit.offset` in llvm), which is resolved at compile time by [GlobalOffset](https://github.com/intel/llvm/blob/sycl/llvm/include/llvm/SYCLLowerIR/GlobalOffset.h) pass. The header file for the pass says that the intrinsic is exposed to SYCL, but libclc is written in OpenCL.

Do I overcomplicate things and can use standard OpenCL built-ins in libclc to implement SYCL functions?

@Naghasan, FYI.

Naghasan replied:
you are a bit @bader , the intrinsics is exposed indirectly and offset is handled in libclc for nvptx and amd.

I think someone pointed out already you can't use `__spirv_GlobalLinearId` for SYCL as I can't find the code in the current version.

maarquitos14 commented on sycl/include/sycl/nd_item.hpp:
```diff
@@ -38,6 +39,273 @@ namespace ext::oneapi::experimental {
+#ifdef __SYCL_DEVICE_ONLY__
+    return __spirv::initGlobalInvocationId<Dimensions, id<Dimensions>>();
+#else
+    // TODO: what should be the implementation on the host? nd_item can be unsed
```
```suggestion
    // TODO: what should be the implementation on the host? nd_item can be used
```

bader commented on sycl/include/sycl/nd_item.hpp:
```diff
@@ -38,6 +39,273 @@ namespace ext::oneapi::experimental {
+  nd_item &operator=(const nd_item &rhs) = default;
+  nd_item &operator=(nd_item &&rhs) = default;
+
+  bool operator==(const nd_item &) const { return true; }
```
This is an important question because it can kill the whole idea of nd_item optimization.
My initial thought was that `nd_item` object can be created only by the runtime and it's not modifiable. In other words, we can consider it to be singleton object for the work-item.
On the other hand, I can see that `operator==` might need to return `false` if `nd_item` argument is created by another kernel instance (i.e. work-item). Now the question is: can we get access to `nd_item` object from another work-item?
I can imagine that the only option is to pass it though the memory shared among work-items (i.e. `local` or `global`), but I can't come-up with the example where it can be done w/o adding UB.
Is it possible to allocate a shared memory for nd_item object and store/load this object using only a copy constructor?

gmlueck replied:
> but I can't come-up with the example where it can be done w/o adding UB.

I think this can be done by using placement-new and the copy constructor.  The application can allocate uninitialized bytes in shared memory and then use placement-new to copy-construct an `nd_item` object into that memory.

Once the object is in shared memory, it could be accessed by other work-items in the kernel execution.

bader replied:
> The application can allocate uninitialized bytes in shared memory and then use placement-new to copy-construct an `nd_item` object into that memory.

Does this involve type casting (from "uninitialized bytes" to "nd_item")? If so, is this casting properly specified? At least "uninitialized bytes" must be properly aligned.

gmlueck replied:
Placement new is a standard C++ feature, and it does not require any type casting.  It is true that the "uninitialized bytes" need to be properly aligned.  You can see more in the [cppreference description](https://en.cppreference.com/w/cpp/language/new#Placement_new).

I admit this is an unusual code pattern, and I don't know if it is something that any real application does.  However, I think it is legal according to the current wording of the spec.

Pennycook replied:
I think this is something that we need to revisit and clarify in the SYCL specification.

Between SYCL 1.2.1 and SYCL 2020, the definition of the `group` class changed to make it clear that `group` (and `sub_group`) are stateless.  I'm having trouble tracking down the commit (and I suspect it's internal to Khronos anyway) but you can see that the 2020 definitions say things like "Return a SYCL `id` representing **the calling work-item’s position** within the `work-group`" (emphasis mine).

Does `nd_item` only have an `operator==` because of Section 4.5.3?  This whole section seems weird to me, now: firstly, because most of these classes are difficult to pass between work-items; and secondly, because the equality operator seems underspecified.  The specification says:

> Equality between two instances of `T` (i.e. `a == b`) must be true if the value of all members are equal

...but the specification for `nd_item` doesn't say what the members are, or if there even are any.

bader replied:
> Does `nd_item` only have an `operator==` because of Section 4.5.3? This whole section seems weird to me, now: firstly, because most of these classes are difficult to pass between work-items; and secondly, because the equality operator seems underspecified. The specification says:
> 
> > Equality between two instances of `T` (i.e. `a == b`) must be true if the value of all members are equal
> 
> ...but the specification for `nd_item` doesn't say what the members are, or if there even are any.

That's exactly right. I think SYCL spec should allow "member-less" implementation of `nd_item` like I tried to prototype in this PR. 
I'm almost 100% sure that no users who passes `nd_item` objects between work-items, but Greg might be right that there is a loophole allowing this use case technically even if has little sense in practice.

gmlueck replied:
>I think this is something that we need to revisit and clarify in the SYCL specification

I created this issue in the SYCL spec repo: KhronosGroup/SYCL-Docs#532.

maarquitos14 replied:
Following up on this discussion: if the implementation is "member-less", does it make sense to have equality/inequality comparison at all? I don't feel very comfortable with a `==` method that always returns `true`, but I can live with it.

AlexeySachkov replied:
With the current wording of the spec, that class has to have `operator==`, because it follows common by-value semantics. I would expect that if we agree that `nd_item` represents a calling work-item, then it doesn't make a lot of sense to have `operator==` in it and therefore the spec will be updated to reflect that.

bader replied:
I'll do follow-up updates if spec changes. Meanwhile, let's keep the operator.

bader commented on sycl/include/sycl/nd_item.hpp:
```diff
@@ -38,6 +39,273 @@ namespace ext::oneapi::experimental {
+  sub_group get_sub_group() const { return sub_group(); }
+
+  size_t __SYCL_ALWAYS_INLINE get_group(int Dimension) const {
+    // Why Size? It returns group id!!!
```
This code is copied from the existing `nd_item` implementation.
If we change it, we should update both places.

bader replied:
No. The comment applies to the variable name on the next line.

bader replied:
I updated the name for both implementations.

bader commented on sycl/include/sycl/nd_item.hpp:
```diff
@@ -38,6 +39,273 @@ namespace ext::oneapi::experimental {
 template <int Dimensions> class root_group;
 }
 
+#if __INTEL_PREVIEW_BREAKING_CHANGES || 1
```
I made a separate commit https://github.com/intel/llvm/pull/12236/commits/59b30d693b3bc228aa7a44a0ea1a19640e67e667 to test new `nd_item` implementation. I'm going to revert this commit before the merge.

bader replied:
I reverted that change.

bader commented on sycl/include/sycl/nd_item.hpp:
```diff
@@ -38,6 +39,273 @@ namespace ext::oneapi::experimental {
+                                    accessMode == access::mode::read_write,
+                                access::fence_space>
+          accessSpace = access::fence_space::global_and_local) const {
+    uint32_t flags = detail::getSPIRVMemorySemanticsMask(accessSpace);
```
@maarquitos14, note, `mem_fence` is handle differently from other `group` methods.
I inlined the implementation, rather than making `group` class method static. It's because it's much simpler than `async_*` and `wait` methods.
On top of that, I pass `accessSpace` argument to SPIR-V built-in, whereas current `nd_item` implementation ignores it. The argument is ignored in `nd_item` method, but `TODO` comment below is from `group` class member implementation. It looks to me that the comment from `group` method tries to explain the code from `nd_item` method.
Are there any plans to clean things up? I'm not sure if I also should ignore `accessSpace` argument or not.

maarquitos14 replied:
Firstly, sorry for the delay, I have been off. 

Honestly, I'm not familiar with this issue, but my understanding of the comment is that it doesn't matter what we pass to SPIR-V because SPIR-V can't easily discern between load-only or store-only operations. 

 Let's see if other folks from @intel/llvm-reviewers-runtime can help here.

bader commented:
> 3\. Replaced global and local liear ID computation with SPIR-V
>     intrinsics.

I removed this part from the patch due to failures in the pre-commit. I need time to investigate the reason for failures, but I don't want to block this PR. This optimization can go in a separate PR.

rolandschulz commented on sycl/include/sycl/group.hpp:
```diff
@@ -353,12 +354,13 @@ template <int Dimensions = 1> class __SYCL_TYPE(group) group {
-  std::enable_if_t<!detail::is_bool<DestDataT>::value &&
-                       std::is_same_v<std::remove_const_t<SrcDataT>, DestDataT>,
-                   device_event>
+  static std::enable_if_t<
```
This is an ABI change. Which makes it not spec compliant. Can we make group stateless too so that this isn't needed?

bader replied:
> Can we make group stateless too so that this isn't needed?

I'm not sure about that, but I'll revert group class changes.

> This is an ABI change.

Are these methods part of the SYCL runtime ABI?

rolandschulz replied:
> Are these methods part of the SYCL runtime ABI?

Sorry. I meant to say API. The spec doesn't define an ABI.
To be specific: The SYCL spec says that the methods aren't static. And it is an user-observable difference whether they are or aren't static.

bader replied:
Reverted.

gmlueck commented on sycl/include/sycl/nd_item.hpp:
```diff
@@ -38,6 +39,273 @@ namespace ext::oneapi::experimental {
+#ifdef __SYCL_DEVICE_ONLY__
+    return __spirv::initGlobalInvocationId<Dimensions, id<Dimensions>>();
+#else
+    // TODO: what should be the implementation on the host? nd_item can be unsed
```
I suspect these TODO's are left over from when we had the host device.

bader replied:
No. I added these TODOs with this patch.
I coped original `nd_item` implementation and modified existing methods. You can see that existing implementation of this method doesn't have this comment.

bader replied:
Done.

gmlueck reviewed (commented):
I'm not sure about this change.  I think a fundamental assumption is that a work-item will never access an `nd_item` object that comes from a different work-item.  I'm not sure that's a valid assumption.

If this *is* really what we intended in the SYCL spec, why aren't all the member functions on `nd_item` declared `static`?

rolandschulz commented:
> I'm not sure about this change. I think a fundamental assumption is that a work-item will never access an `nd_item` object that comes from a different work-item. I'm not sure that's a valid assumption.

I think it is a valid assumption. But I agree that the spec isn't clear enough that it is invalid to access an `nd_item` from a different work-item. And I thought there was actual a Khronos issue open for that. But I can't find it. I do recall this was being discussed before.

> If this _is_ really what we intended in the SYCL spec, why aren't all the member functions on `nd_item` declared `static`?

Because either implementation should be made possible by the spec.

bader commented:
> > I'm not sure about this change. I think a fundamental assumption is that a work-item will never access an `nd_item` object that comes from a different work-item. I'm not sure that's a valid assumption.
> 
> I think it is a valid assumption. But I agree that the spec isn't clear enough that it is invalid to access an `nd_item` from a different work-item. And I thought there was actual a Khronos issue open for that. But I can't find it. I do recall this was being discussed before.
> 

I would like this assumption to be valid to be able to optimize existing applications without changing their code. 
If it's not valid, I think we need to add new API providing fine-grained access to `nd_item` data. For instance, if the application uses only local size along z dimension, it should be able to get just this value and do not create the whole `nd_item` state.

Pennycook commented on sycl/include/sycl/nd_item.hpp:
```diff
@@ -38,6 +41,492 @@ namespace ext::oneapi::experimental {
+
+  nd_range<Dimensions> get_nd_range() const { return nd_range<Dimensions>(); }
+
+  void barrier(access::fence_space accessSpace =
+                   access::fence_space::global_and_local) const {
+    uint32_t flags = _V1::detail::getSPIRVMemorySemanticsMask(accessSpace);
+    __spirv_ControlBarrier(__spv::Scope::Workgroup, __spv::Scope::Workgroup,
+                           flags);
+  }
```
Should this include some sort of deprecation notice?  SYCL 2020 removed this function completely, in favor of `sycl::group_barrier`.

bader replied:
If so, we need to change stateful implementation too. As it's not directly related to making `nd_item` stateless, I created https://github.com/intel/llvm/issues/12531 to make this change separately.
@intel/sycl-language-enabling-triage, FYI.

AlexeySachkov replied:
Yes, we have an internal task/bug report for the deprecation and removal of the method. We have tried it once already (even just deprecation), but reverted the change under pressure from some customers. The problem with `group_barrier` is that it doesn't accept fence space argument and uses the widest scope possible for memory fence. This causes performance regressions on some of our HW.

We should have a bug reported against corresponding low-level runtimes, but I'm not sure if there has been any meaningful progress on it.

AlexeySachkov replied:
Those are two different kind of scopes/spaces:

[`OpControlBarrier`](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpControlBarrier) accepts 3 arguments:
1. execution scope - which work-items should wait for other work-items
2. memory scope - how "wide" mem-fence should be: do we care about synchronizing memory operations what may have happened in other work-groups, or just this one is fine?
3. memory semantics - address spaces to synchronize + memory order (acquire, release, etc.)

[`access::fence_space`](https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/access/access.hpp#L43) is essentially an address space, which needs to be synchronized and it is a legacy from SYCL 1.2.1. As you can see on the snippet above, `nd_item::barrier` does execution synchronization of a work-group and mem-fence within a work-group for *the specified address space*.

[`group_barrier`](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_group_barrier) accepts `memory_scope`, which is how wide the mem fence should be, but not which address spaces it should cover.

With `group_barrier` we end up doing execution synchronization of a work-group and mem-fence within a specified scope for *all address spaces*. See [1](https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/group_barrier.hpp), [2](https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/detail/spirv.hpp#L1016)

`group_barrier` leads to emission of some extra mem fences and if I understand correctly, there is a bug/missing optimization in IGC that does not remove those extra fences, which leads to extra synchronization and performance degradation. Since `group_barrier` does not allow to pass an address space, users have no workaround (except for using the legacy `nd_item::barrier`) for that bug.

bader replied:
@AlexeySachkov, are you saying that to fix that bug we need to extend `group_barrier` API? Could you open a separate issue for that, please?

AlexeySachkov replied:
> @AlexeySachkov, are you saying that to fix that bug we need to extend `group_barrier` API? Could you open a separate issue for that, please?

SYCL 2020 version of the barrier definitely has less capabilities than SYCL 1.2.1 `nd_item::barrier` and that puts an extra burden on the implementation to maintain the same level of performance. However, as I understand it, that was a deliberate choice in the spec direction to have a single happens-before relationship for all address spaces at once to simplify user experience with SYCL. It comes with a cost, unless mitigated by optimizations made at different levels of SW stack.

@Pennycook was the one who led that effort,  I believe. See internal Khronos PR 520 (removes mentions of address spaces from `group_barrier`), issue 354 (prompted PR 520) and PR 420 (first (?) mention of single happens-before relationship for all address spaces).

Pennycook replied:
Yes, this was deliberate.  There were lots of reasons, but my favorite example is that the combination of C++ templates, generic address spaces and a split happens-before makes it _really_ easy to write code with hard-to-find bugs in it:
```c++
template <typename T1, typename T2>
void buggy(nd_item<1> it, T1* ptr1, T2* ptr2) {
  if (it.get_global_id(0)) ptr1 = x;
  if (it.get_global_id(0)) ptr2 = y;
  nd_item::barrier(access::global_space); // oops
  if (ptr1 == x && ptr2 == y) {
    /* do something */
  }
}
```

This code will work if both pointers happen to point to the global address space, but may fail if either pointer points to something else.  Detecting which space to fence statically would require the user to opt-in to using address space qualifiers (which they typically don't want to do).  Detecting which space to fence dynamically would require runtime checks of certain pointers.  Calling `nd_item::barrier()` to fence both spaces is the easy fix, and that's equivalent to calling `group_barrier`.

There may be some special corner cases where somebody who knows what they're doing has a reason to generate a fence for one address space.  But my view is that if a developer is capable of fully understanding the ramifications of a split-happens before on a specific architecture, they should also be capable of calling something like inline assembly or SPIR-V intrinsic.

maarquitos14 commented on sycl/include/sycl/nd_item.hpp:
```diff
@@ -38,6 +41,492 @@ namespace ext::oneapi::experimental {
+#ifdef __SYCL_DEVICE_ONLY__
+    return __spirv::initGlobalInvocationId<Dimensions, id<Dimensions>>();
+#else
+    assert(false && "nd_item methods can't be used on the host!");
```
Should we throw instead? Currently, a user running non-debug code would just receive the wrong result without any error/warning, and it might be confusing. 

Same for other locations were we do something similar.

jzc replied:
what about something like this? This compiles; whether it should/has defined behavior is unclear - should nd_item be device copyable?

```c++
#include <sycl/sycl.hpp>

int main() {
  using namespace sycl;
  static_assert(is_device_copyable_v<nd_item<1>>);
  static_assert(is_device_copyable_v<nd_item<2>>);
  static_assert(is_device_copyable_v<nd_item<3>>);
  queue q;
  {
    auto *p = malloc_shared<nd_item<1>>(1, q);
    q.parallel_for(nd_range<1>(8, 4), [=](nd_item<1> it) {
      if (it.get_global_id(0) == 5)
        *p = it;
    }).wait();
    if (p->get_global_id(0) != 5) std::cout << "fail\n";
    free(p, q);
  }
  {
    buffer<nd_item<1>, 1> buf { 1 };
    q.submit([&](handler &cgh) {
      accessor acc { buf, cgh };
      cgh.parallel_for(nd_range<1>(8,4), [=](nd_item<1> it) {
        if (it.get_global_id(0) == 5)
          acc[0] = it;
      });
    });
    host_accessor ha { buf };
    if (ha[0].get_global_id(0) != 5) std::cout << "fail\n";
  }
}
```

bader replied:
This should not be valid. The spec doesn't say it explicitly, but `nd_item` should not be available on the host.

https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_class_availability section says that it must be available in SYCL kernels.
https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:byval-semantics assumes that some of these types are not available on the host.

`nd_item` must be one the classes which are not available on the host. `nd_item` methods have meaning only in SYCL kernel scope.

@intel/dpcpp-specification-reviewers, please, follow-up on spec clarifications if needed.

maarquitos14 replied:
> `nd_item` must be one the classes which are not available on the host. `nd_item` methods have meaning only in SYCL kernel scope.

Exactly, totally agreed. That's why I think throwing would clearly state that this is not an implementation error, but a user error, if anybody is not very familiar with SYCL and tries to do something like Justin's example.

Anyway, this is not blocking and I can accept either way.

bader replied:
Done in https://github.com/intel/llvm/pull/12236/commits/051ce6c68b9d4f5441eff36f9d61fbe7b20e6ae0.

maarquitos14 commented on sycl/include/sycl/nd_item.hpp:
```diff
@@ -38,6 +41,492 @@ namespace ext::oneapi::experimental {
+#endif
+  }
+
+  nd_range<Dimensions> get_nd_range() const { return nd_range<Dimensions>(); }
```
Previous implementation passes arguments to `nd_range`. Don't we need them here?

bader replied:
Nice catch. I think I just forgot to implement this method.
The scary thing is that pre-commit passed! We have 310 end-to-end tests using `nd_item` and none of them test this method.

bader replied:
Make sense to me. I've filed https://github.com/intel/llvm/issues/12563 to follow-up on that.

maarquitos14 reviewed (approved):
Just a couple of non-blocking issues left. LGTM.

ldrumm commented:
Do you have benchmarks for before and after?
What about binary sizes?

AerialMantis commented on sycl/include/sycl/nd_item.hpp:
```diff
@@ -38,6 +41,502 @@ namespace ext::oneapi::experimental {
+    return Id;
+  }
+
+  size_t __SYCL_ALWAYS_INLINE get_global_linear_id() const {
```
I wonder whether there is a potential performance impact here for code which calls `get_global_linear_id()` quite liberally, how does the code generated for this implementation compare with the original implementation?

bader replied:
They match. 
As I removed the `item` member, I just inlined `item::get_linear_id` implementation. 

https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/detail/item_base.hpp#L33-L42

I don't expect any performance changes in this regard.

You might find [this discussion](https://github.com/intel/llvm/pull/12236#discussion_r1434590193) interesting. I tried to use SPIR-V intrinsic to get liner ID, but it broke some precommit tests and I wasn't sure how to handle offsets on NVIDIA and AMD GPUs.

AerialMantis reviewed (approved):
This change LGTM, just one comment regarding `get_global_linear_id()`.

Following on from @gmlueck's comment, I believe that it was the intent of the SYCL specification that an `item`/`nd_item` represent a single work-item, and should not be accessed outwith the work-item it was created in, to allow an implementation freedoms such as this. I think we may need a clarification in the specification to make this more explicit.

bader commented:
> Do you have benchmarks for before and after?

Nothing interesting I can share. I suppose the tests I tried a quite small and compiler is able to optimize existing implementation well enough to eliminate the difference in performance.

> What about binary sizes?

Could you provide more context for the question, please? Are you asking about the size of the GPU (other) code? Are you concerned that calling compiler intrinsics will increase the code size?
I personally don't think so, but I'm not sure if we have any benchmarks which measure GPU code size.

bader commented:
@gmlueck, ping.

gmlueck commented on sycl/include/sycl/nd_item.hpp:
```diff
@@ -38,6 +41,502 @@ namespace ext::oneapi::experimental {
+    return Id;
+  }
+
+  __SYCL2020_DEPRECATED("offsets are deprecated in SYCL 2020")
```
I wonder if it makes sense to have deprecated functions under the `__INTEL_PREVIEW_BREAKING_CHANGES` guard.  Won't these functions be removed at the next major release?  If that is the case, I think it might make sense to omit the deprecated functions in this mode because that would be a more accurate preview of the next major release.

bader replied:
I suggest removing deprecated methods in a separate PR. If there are going to be removed with in next major release, we should put existing deprecated methods of `nd_item` under `#ifndef __INTEL_PREVIEW_BREAKING_CHANGES`. Otherwise, we will have different sets of methods for stateless and stateful implementations.

bader commented:
Post-commit validation results:
Arc:

Failed Tests (1):
  SYCL :: Reduction/reduction_range_3d_rw_reducer_skip.cpp

This failure has seen in other post-commit runs.
